### PR TITLE
Implement subresource for Node.Spec.ConfigSource

### PR DESCRIPTION
**Pull request by mtaufen**
This PR implements a "nodes/configsource" subresource, which must be
used to update Node.Spec.ConfigSource instead of a standard Node update.

This allows us to write separate access control rules for managing
Node.Spec.ConfigSource vs updating other parts of the Node object.

See: https://github.com/kubernetes/kubernetes/issues/58362

```release-note
Updates to Node.Spec.ConfigSource should now be performed via the "nodes/configsource" subresource.
```

Please sanity check this, it's the first time I've implemented a subresource and I'm not sure if this is the right way to do it (just following in the footsteps of preexisting code).

### CODE REVIEW

File: staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_node.go [hunk 1]
```diff
@@ -127,3 +127,13 @@ func (c *FakeNodes) Patch(name string, pt types.PatchType, data []byte, subresou
 	}
 	return obj.(*core_v1.Node), err
 }
+
+// UpdateConfigSource takes the representation of a node and updates it. Returns the server's representation of the node, and an error, if there is any.
+func (c *FakeNodes) UpdateConfigSource(nodeName string, node *core_v1.Node) (result *core_v1.Node, err error) {
+	obj, err := c.Fake.
+		Invokes(testing.NewRootUpdateAction(nodesResource, node), &core_v1.Node{})
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_node.go]: Was this autogenerated? It seems like this should use a NewUpdateSubresourceAction.

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -4042,6 +4045,20 @@ func ValidateNodeUpdate(node, oldNode *core.Node) field.ErrorList {
...
+	if source.ConfigMapRef != nil {
+		count++
+	}
+	// add more subfields here in the future as they are added to NodeConfigSource
+
+	// exactly one reference subfield must be non-nil
+	if count != 1 {
+		return field.ErrorList{field.Invalid(fldPath, source, "exactly one reference subfield must be non-nil")}
+	}
+	return field.ErrorList{}
```
Comment (mikedanese) [inline pkg/apis/core/validation/validation.go]: nit: convention in validation functions is to create an empty allErrs at the start of a function, append validation errors to it, and return it at the end

File: pkg/registry/core/node/storage/storage.go [hunk 3]
```diff
@@ -72,6 +73,28 @@ func (r *StatusREST) Update(ctx genericapirequest.Context, name string, objInfo
...
+var (
+	_ = rest.Getter(&ConfigSourceREST{})
+	_ = rest.Updater(&ConfigSourceREST{})
+)
+
+func (r *ConfigSourceREST) New() runtime.Object {
+	return &api.Node{}
+}
+
+func (r *ConfigSourceREST) Get(ctx genericapirequest.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
```
Comment (mikedanese) [inline pkg/registry/core/node/storage/storage.go]: I would go ahead and leave this out unless you have a need for it. It's not accessible from the client as is.

File: staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_node.go (hunk 1 above)
Comment (mtaufen) [reply staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_node.go]: Yes, autogenerated. @caesarxuchao

File: pkg/registry/core/node/storage/storage.go (hunk 3 above)
Comment (mtaufen) [reply pkg/registry/core/node/storage/storage.go]: sounds good

File: pkg/apis/core/validation/validation.go (hunk 2 above)
Comment (mtaufen) [reply pkg/apis/core/validation/validation.go]: done

File: staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_node.go (hunk 1 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/client-go/kubernetes/typed/core/v1/fake/fake_node.go]: That seems wrong. cc @mfojtik.

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -4036,6 +4039,21 @@ func ValidateNodeUpdate(node, oldNode *core.Node) field.ErrorList {
 	return allErrs
 }
 
+func validateNodeConfigSource(source *core.NodeConfigSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	count := int(0)
+	if source.ConfigMapRef != nil {
+		count++
```
Comment (mtaufen) [inline pkg/apis/core/validation/validation.go]: TODO: also check ConfigMapRef name, namespace, uid are non-empty here

**Comment by jberkus** [comment]
we are now in code slush, so please add status/approved-for-milestone when this PR is ready to merge.  thanks!

@mtaufen

**Comment by mtaufen** [comment]
rebased

**Pull request closed without merging**